drivers/timer/npcx_itim_timer.c

```diff
@@ -0,0 +1,319 @@
+ * - Its prescaler is set to 1 and provide the kernel cycles reading without
+ *   handling overflow mechanism.
+ * - After ec entered "sleep/deep sleep" power state which is used for better
+ *   power consumption, then its clock will stop.
+ *
+ * - A event timer based on an ITIM32 (Internal 32-bit timer) instance, clocked
+ *   by LFCLK which frequency is 32KHz and still activated when ec entered
+ *   "sleep/deep sleep" power state.
+ * - Provide a system clock timeout notification. In its ISR, the driver informs
+ *   the kernel that the specified number of ticks have elapsed.
+ * - Its prescaler is set to 1 and the formula between event timer's cycles and
+ *   ticks is 'cycles = (ticks * 32768) / CONFIG_SYS_CLOCK_TICKS_PER_SEC'
+ * - Compensate reading of ITIM64 which clock is gating after ec entered
+ *   "sleep/deep sleep" power state if CONFIG_PM is enabled.
+ */
+
+#include <drivers/clock_control.h>
+#include <drivers/timer/system_timer.h>
+#include <sys_clock.h>
+#include <spinlock.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(itim, LOG_LEVEL_ERR);
+
+#define NPCX_ITIM32_MAX_CNT 0xffffffff
+#define NPCX_ITIM64_MAX_HALF_CNT 0xffffffff
+#define EVT_CYCLES_PER_SEC LFCLK /* 32768 Hz */
+#define SYS_CYCLES_PER_TICK (sys_clock_hw_cycles_per_sec() \
+					/ CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define EVT_CYCLES_FROM_TICKS(ticks) \
+	ceiling_fraction(ticks * EVT_CYCLES_PER_SEC, \
+			 CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+/* Instance of system and event timers */
+static struct itim64_reg *const sys_tmr = (struct itim64_reg *)
+					DT_INST_REG_ADDR_BY_NAME(0, sys_itim);
+static struct itim32_reg *const evt_tmr = (struct itim32_reg *)
+					DT_INST_REG_ADDR_BY_NAME(0, evt_itim);
+
+static const struct npcx_clk_cfg itim_clk_cfg[] = NPCX_DT_CLK_CFG_ITEMS_LIST(0);
+
+static struct k_spinlock lock;
+/* Announced cycles in system timer before executing z_clock_announce() */
+static uint64_t cyc_sys_announced;
+/* Current target cycles of time-out signal in event timer */
+static uint32_t cyc_evt_timeout;
+#if defined(CONFIG_PM)
+/* Total cycles of system timer stopped in "sleep/deep sleep" mode */
+static uint64_t cyc_sys_compensated;
+/* Current cycles in event timer when ec entered "sleep/deep sleep" mode */
+static uint32_t cyc_evt_enter_deep_idle;
+#endif
+
+/* ITIM local inline functions */
+static inline uint64_t npcx_itim_get_sys_cyc64(void)
+{
+	uint32_t cnt64h, cnt64h_check, cnt64l;
+
+	/* Read 64-bit counter value from two 32-bit registers */
+	do {
+		cnt64h_check = sys_tmr->ITCNT64H;
+		cnt64l  = sys_tmr->ITCNT64L;
+		cnt64h = sys_tmr->ITCNT64H;
+	} while (cnt64h != cnt64h_check);
+
+	cnt64h = NPCX_ITIM64_MAX_HALF_CNT - cnt64h;
+	cnt64l = NPCX_ITIM64_MAX_HALF_CNT - cnt64l + 1;
+
+	/* Return current value of 64-bit counter value of system timer */
+#if defined(CONFIG_PM)
+	return ((((uint64_t)cnt64h) << 32) | cnt64l) + cyc_sys_compensated;
+#else
+	return (((uint64_t)cnt64h) << 32) | cnt64l;
+#endif
```

Comment:
Here too, prefer `IS_ENABLED` for readability.

Reply:
Replacing this with IS_ENABLED() macro will cause a compiler error since 'cyc_sys_compensated' is undeclared if CONFIG_PM is not enabled. If I remove line 69 & 74, we need to move 'cyc_evt_enter_deep_idle' variable to line 309 in case 'unused variable' warning. Does it improve readability if I do so?

Reply:
I see, up to you, personally I prefer to set line 71 to:
```
__maybe_unused static uint64_t cyc_sys_compensated;
```

That avoids the warning. Though looking at the code further I think you can just always declare cyc_sys_compensated which will default to 0, then you don't even need this if statement here and you can just always `+ cyc_sys_compensated`. The compiler should have an easy enough time figuring out that `cyc_sys_compensated` is always 0 and replace it accordingly.

* I recognize that I'm always in favor of offloading work to the compiler/optimizer in favor of simpler code, if you disagree I'm not going to block on this, just throwing my 2 cents.

Reply:
_maybe_unused seems doesn't work on Zephyr. But I use "__unused" instead of it. I agree with you that " offloading work to the compiler/optimizer in favor of simpler code" improves readability. Thanks for your comment. I learn a lot.

Reply:
Done.